### PR TITLE
Convert rpc to request/response

### PR DESCRIPTION
**Pull request by kerzhner**
This PR refactors the network stack of the rpc code. In this refactor, an connection is an interface with [Request and Subscribe methods](https://github.com/statechannels/go-nitro/blob/f19d23b8c8def1c3471bed46dc6802078fcb6a65/network/transport/connection.go#L3). An rpc client sends json rpc requests over the network. An rpc server subscribes to json rpc requests. Currently, there is one implementation of a connection: nats. We should try adding another connection type (such as websockets) in the near future to ensure that nats is sufficiently abstracted out of the rpc client and server.

Note that jsonrpc error handling logic has been removed. The logic was untested. https://github.com/statechannels/go-nitro/issues/1103 has been filed to reimpliment the jsonrpc error handling.

### CODE REVIEW

File: network/transport/nats/nats_connection.go [hunk 1]
```diff
@@ -1,73 +1,50 @@
...
+	if msg == nil {
+		return nil, fmt.Errorf("received nill data for request %v with error %w", topic, err)
 	}
-	log.Trace().Msgf("published message on %s.\ndata: %v", t, string(data))
+	return msg.Data, err
 }
 
-func (c *natsConnection) Recv() ([]byte, error) {
-	msg := <-c.msgChannel
+func (c *natsConnection) Subscribe(topic serde.RequestMethod, handler func([]byte) []byte) error {
```
Comment (lalexgap) [inline network/transport/nats/nats_connection.go:34]: nit, godoc

File: network/transport/nats/nats_connection.go [hunk 2]
```diff
@@ -1,73 +1,50 @@
...
-	}
-
-	return sub, nil
-}
-
-func (c *natsConnection) Send(t string, data []byte) {
-	err := c.nc.Publish(t, data)
-	if err != nil {
-		log.Error().Err(err).Msgf("failed to send message on topic: %s. msg: %s", t, string(data))
+func (c *natsConnection) Request(topic serde.RequestMethod, data []byte) ([]byte, error) {
```
Comment (lalexgap) [inline network/transport/nats/nats_connection.go:26]: nit, godoc

File: network/client.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	"github.com/statechannels/go-nitro/network/serde"
+	"github.com/statechannels/go-nitro/network/transport"
+	"github.com/statechannels/go-nitro/protocols"
+	"github.com/statechannels/go-nitro/protocols/directdefund"
+	"github.com/statechannels/go-nitro/protocols/directfund"
+	"github.com/statechannels/go-nitro/protocols/virtualdefund"
+	"github.com/statechannels/go-nitro/protocols/virtualfund"
+)
+
+type Response struct {
```
Comment (lalexgap) [inline network/client.go:18]: Nit but we could add some godoc to this file

**Review by lalexgap — APPROVED** [review]
This looks like it simplifies things a lot! 🎉

**Comment by lalexgap** [comment]
I've rebased this on https://github.com/statechannels/go-nitro/pull/1104 to hopefully resolve the flicker

**Comment by lalexgap** [comment]
> I've rebased this on #1104 to hopefully resolve the flicker

The first run [failed](https://github.com/statechannels/go-nitro/actions/runs/4110829517/attempts/1) but it looks like it's a timeout error and not the original flicker we were seeing. Rerunning succeeded.

File: network/transport/nats/nats_connection.go (hunk 2 above)
Comment (kerzhner) [reply network/transport/nats/nats_connection.go:29]: [b2c9f63](https://github.com/statechannels/go-nitro/pull/1102/commits/b2c9f63b3d04ac5c1dd8ed306883b2b887728d69)

File: network/client.go (hunk 3 above)
Comment (kerzhner) [reply network/client.go:18]: [b2c9f63](https://github.com/statechannels/go-nitro/pull/1102/commits/b2c9f63b3d04ac5c1dd8ed306883b2b887728d69)

File: network/transport/nats/nats_connection.go (hunk 1 above)
Comment (kerzhner) [reply network/transport/nats/nats_connection.go:40]: Good call! Done in [b2c9f63](https://github.com/statechannels/go-nitro/pull/1102/commits/b2c9f63b3d04ac5c1dd8ed306883b2b887728d69)

File: network/client.go (hunk 3 above)
Comment (kerzhner) [reply network/client.go:20]: Good call. Added in [20f6122](https://github.com/statechannels/go-nitro/pull/1102/commits/20f61224c8b7d1032c40a5ba6ec910bf09be4b96).

File: client/rpc/client.go [hunk 4]
```diff
@@ -62,190 +49,57 @@ func (rc *RpcClient) CreateVirtual(intermediaries []types.Address, counterparty
...
-	})
-}
-
-// handleResponse handles a response from the rpc server for the given client
-// It is not a member of the RpcClient so it can take advantage of generics
-func handleResponse[T serde.ResponsePayload](rc *RpcClient, data []byte) {
-	rpcResponse := serde.JsonRpcResponse[T]{}
-	err := json.Unmarshal(data, &rpcResponse)
+func waitForRequest[T serde.RequestPayload, U serde.ResponsePayload](rc *RpcClient, requestData T) U {
+	resChan, err := network.Request[T, U](rc.connection, requestData, rc.logger)
```
Comment (geoknee) [inline client/rpc/client.go:94]: Nice! 🚀

**Review by geoknee — APPROVED** [review]
Not reviewed deeply, but this looks pretty good to me.

**Pull request merged**